src/vs/workbench/contrib/debug/browser/disassemblyView.ts

```diff
@@ -0,0 +1,314 @@
+	icon: HTMLElement
+}
+
+class BreakpointRenderer implements ITableRenderer<IDisassembledInstructionEntry, IBreakpointColumnTemplateData> {
+
+	static readonly TEMPLATE_ID = 'breakpoint';
+
+	templateId: string = BreakpointRenderer.TEMPLATE_ID;
+
+	renderTemplate(container: HTMLElement): IBreakpointColumnTemplateData {
+		const icon = append(container, $('.disassembly-view'));
+		icon.classList.add('codicon');
+
+		icon.style.display = 'flex';
+		icon.style.alignItems = 'center';
+		icon.style.justifyContent = 'center';
+
+		return { container, icon };
+	}
+
+	renderElement(element: IDisassembledInstructionEntry, index: number, templateData: IBreakpointColumnTemplateData, height: number | undefined): void {
+		// TODO: stepping
+		// const debugStackframe = 'codicon-' + icons.debugStackframe.id;
+		// const debugStackframeFocused = 'codicon-' + icons.debugStackframeFocused.id;
+
+		// TODO: see getBreakpointMessageAndIcon in vs\workbench\contrib\debug\browser\breakpointEditorContribution.ts
+		//       for more types of breakpoint icons
+		if (element.allowBreakpoint) {
+			const breakpointIcon = 'codicon-' + icons.breakpoint.regular.id;
+			const breakpointHintIcon = 'codicon-' + icons.debugBreakpointHint.id;
+
+			if (element.isBreakpointSet) {
+				templateData.icon.classList.add(breakpointIcon);
+			}
+
+			templateData.container.onmouseover = () => {
+				templateData.icon.classList.add(breakpointHintIcon);
+			};
+
+			templateData.container.onmouseout = () => {
+				templateData.icon.classList.remove(breakpointHintIcon);
+			};
+
+			templateData.container.onclick = () => {
+				if (element.isBreakpointSet) {
+					element.isBreakpointSet = false;
+					templateData.icon.classList.remove(breakpointIcon);
+				} else if (element.allowBreakpoint) {
+					element.isBreakpointSet = true;
+					templateData.icon.classList.add(breakpointIcon);
+				}
+			};
+		}
+	}
+
+	disposeTemplate(templateData: IBreakpointColumnTemplateData): void {
+		templateData.container.onclick = null;
+		templateData.container.onmouseover = null;
+		templateData.container.onmouseout = null;
+	}
+
+}
+
+interface IInstructionColumnTemplateData {
+	// TODO: hover widget?
+	instruction: HTMLElement;
+}
+
+class InstructionRenderer implements ITableRenderer<IDisassembledInstructionEntry, IInstructionColumnTemplateData> {
+
+	private static readonly INSTRUCTION_ADDR_MIN_LENGTH = 25;
+	private static readonly INSTRUCTION_BYTES_MIN_LENGTH = 30;
+
+	static readonly TEMPLATE_ID = 'instruction';
+
+	templateId: string = InstructionRenderer.TEMPLATE_ID;
+
+	constructor(private readonly disassemblyView: DisassemblyView) {
+	}
+
+	renderTemplate(container: HTMLElement): IInstructionColumnTemplateData {
+		const instruction = append(container, $('.instruction'));
+		this.applyFontInfo(instruction);
+		return { instruction };
+	}
+
+	renderElement(element: IDisassembledInstructionEntry, index: number, templateData: IInstructionColumnTemplateData, height: number | undefined): void {
```

Comment:
@isidorn Looks like there may be problems rendering the rows if the content is very long. Even when allow horizontal scrolling, the initial width of the table seems to limit the length of a row.

Also, I didn't realize this issue before, but since the first column is breakpoints column, we don't want it disappear when scrolling horizontally. So is there a way to keep it fixed from horizontal scrolling?

Reply:
Table should allow infinite width, you can just scroll to reach it. Can you please clarify what is the problem? You can not reach the end of the row? Are you sure you are properly calling `layout`?

I think it is currently not possible to have rows which are always visible in the table. I am double checking with Joao, but we can think about this as step 2 as an enhancement to the table.
Alternative approach would be to use the list instead of the table and then the whole row rendering is up to you so you can use css tricks to make sure that the breakpoint is always visible

Reply:
I just got back from Joao, and horizontal scrolling is not supported by the table https://github.com/microsoft/vscode/issues/117049

So either we add horizontal scrolling some time in the future
Or we should consider to use the list instead and leave the rendering and positioning of fields to the implementation.

Reply:
Thanks for checking! We think a table should be fine for the first iteration. Anyway, I tried list this afternoon but wasn't able to figure out how to fix an element from scrolling. Looks like the most commonly used CSS style for this is `position: fixed`, but this doesn't work in a list. Any suggestions?

Reply:
@xisui-MSFT this would need some css magic. Your `position:fixed` sounds like a good idea. Though I am not sure if this will work horizontally.
The list should not prevent you from controlling your css layout. So it should be possible, if you ran out of ideas let me know and I can investigate online...

Reply:
Actually I have already run out of ideas... I tried quite a few combinations of css styles but none of them work. Also the behavior seems to be different than when I tried to test with a simple webpage since `position:fixed` does work there. 

Although looks like `position:sticky` may also be a good candidate, it doesn't work either. Feels like the concept of viewport with respect to entries in the list is different from that of the entire page...

Would be great if you could take a look. Thanks!

Reply:
I checked with our css wizard and he said to try to go with `position:sticky` as you mentioned. He linked me this article, maybe it helps https://elad.medium.com/css-position-sticky-how-it-really-works-54cd01dc2d46

Reply:
The only thing that wasn't so obvious that mentioned in this article is it requires a container to work. However, a list entry is already a container, so unfortunately it doesn't really help. Also I try to add another container under the list entry, and still no luck...